prelude/prelude.js

```diff
@@ -16,6 +16,10 @@ declare class $ReadOnlyArray<+T> {
   @@iterator(): Iterator<T>;
 }
 
+type TemplateStringsArray = interface extends $ReadOnlyArray<string> {
```

Comment:
The items of the array can be undefined.

```
$ node
> (x => x)`\u`
[ undefined ]
> (x => x.raw)`\u`
[ '\\u' ]
```

¯\\\_(ツ)_/¯

Reply:
Why would you use tags with `undefined` array

Reply:
Because somebody accidentally put an "invalid escape" in there.

Reply:
Then it would error?

Reply:
Yeah, if the author of the tag expects all items to be string it'll probably crash with `TypeError: Cannot read property 'slice' of undefined` or something like that.

Reply:
I mean typechecker would error

Reply:
Oh, cool, I didn't even think of that the typechecker could statically know if there will be `undefined`s in the array or not!

I tried to see what's it like today on [Try Flow](https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVUCuA7AxgFwEs5swo44AKQ-AUwFsBnALjAEEAnDgQwE8AeRvg6FsAcwB8ASlZCR4sAG9UYMB1r5MHUjQaMAdPW4AHSozABeCWAP44AVWPHaHAMLdGtSlKn6AVnCi3gDcqAC+6ORwAAbcADqY0ahAA) but it gave me a parse error.

```
/* @flow */

function foo(items: Array<string>): string {
  return items.map(s => s.toUpperCase()).join();
}

foo`a\u`
```

```
 foo`a\u`
         ^ Unexpected token ILLEGAL
```	

It would be awesome though if Flow could detect that I'm trying to call `foo` with `Array<string | void>` which is incompatible.

Reply:
How can I get `undefined` without using `\u`?

Reply:
> How can I get `undefined` without using `\u`?

This article seems to list all the cases:

http://2ality.com/2016/09/template-literal-revision.html#problem-some-text-is-illegal-after-backslashes

> TypeScript has same parse error `Hexadecimal digit expected.`

Yeah, maybe they haven't made updates for this ES2018 feature yet. I'll try and see if I can find a github issue about it.